Dockerfile

```diff
@@ -1,19 +1,79 @@
-FROM ruby:3.1.3
+ARG RUBY_VERSION
+FROM ruby:$RUBY_VERSION-slim-bookworm AS base
 
-# Install nodejs on the default ruby 3 image
-RUN curl -sL https://deb.nodesource.com/setup_18.x | bash - && \
-      apt-get install -y nodejs build-essential && \
-      npm install -g pnpm@8.7.4
+RUN test -n "$RUBY_VERSION"
 
 WORKDIR /app
 
-COPY Gemfile ./Gemfile
-COPY Gemfile.lock ./Gemfile.lock
-COPY package.json ./package.json
-COPY pnpm-lock.yaml ./pnpm-lock.yaml
-RUN bundle install
-RUN pnpm install
+# Don't automatically delete packages after installing them. (We want to cache them via Docker.)
+RUN rm /etc/apt/apt.conf.d/docker-clean
 
+# Install base packages
+RUN --mount=type=cache,sharing=private,target=/var/lib/apt/lists \
+  --mount=type=cache,sharing=private,target=/var/cache/apt \
+  apt-get update && \
+  apt-get install --no-install-recommends -y \
+  libjemalloc2 postgresql-client
+
+ENV RAILS_ENV="production"
+ARG GEMS_DIRECTORY=vendor/bundle
+
+# Use jemalloc for memory savings.
+ENV LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so.2
+
+RUN bundle config set path "${GEMS_DIRECTORY}" && \
+  bundle config set without development:test
+
+
+# BEGIN build step: Create throw-away build stage to reduce size of final image
+# \/ \/ \/ \/ \/ \/ \/ \/ \/
+FROM base AS build
+
+# Install packages needed to build gems
+RUN --mount=type=cache,sharing=private,target=/var/lib/apt/lists \
+  --mount=type=cache,sharing=private,target=/var/cache/apt \
+  apt-get update -qq && \
+  apt-get install --no-install-recommends -y \
+  build-essential curl git libpq-dev unzip
+
+# Configure bundler and install application gems
+RUN \
+  bundle config set --local clean 1 && \
+  bundle config set --local deployment 1 && \
+  bundle config set --local path /app/.cache/bundle
+COPY Gemfile Gemfile.lock .ruby-version ./
+RUN --mount=type=cache,sharing=private,target=/app/.cache/bundle \
```

Comment:
Major thanks to this comment for helping me to get this cacheing working: https://github.com/moby/buildkit/issues/ 1173#issuecomment-2242055616 .